### ISSUE TITLE
Magento 2.1 CE JS freezes browser on admin product page

### ISSUE THREAD

**Issue by PingusPepan** [magento/magento2#6413]
I have this issue since Magento 2.1. Magento 2.0 worked fine.

When I try to add a product or edit existing one, the page loads fine and then the Ajax spinner starts spinning. When it reaches the file `vendor/magento/module-ui/view/base/web/templates/grid/filters/elements/ui-select-optgroup.html` (`pub/static/adminhtml/Magento/backend/en_US/Magento_Ui/templates/grid/filters/elements/ui-select-optgroup.html` in browser) it stops spinning and hangs the browser for few minutes. I suspect this is because of categories listing.
We have 11 root categories, each containing multiple categories with subcategories.
I have tried Firefox and Chrome, both with the same result.
I have Intel core i5 and 16GB RAM in my computer and the browser still freezes. People usually working in the admin do not have those machines and they have to kill the browser. For us is the admin completely unusable in current state.
### Preconditions

Magento version 2.1.1 CE (2.1.0 acts the same)
PHP: 7.0.10 with opcache
MySQL: 10.0.20-MariaDB
Session storage: Redis
Cache: Redis
FPC: Redis (no Varnish)

All Magento cache is enabled

`catalog_category_entity` table has **4712** rows.
### Steps to reproduce
1. Prepare >10 root categories
2. Fill with many categories and subcategories
3. Try to add or edit product
### Expected result
1. Not freezing the browser
### Actual result
1. Freezed browser

**Comment by PingusPepan** [comment]
As far as I know, this affects admin only, so maybe the Admin Panel label should be added.

**Comment by philipvandebriel** [comment]
Hi,

We have exact the same problem since Magento 2.1, edit product in admin then page freeze about 1 min. We have about 4000 categories. In magento 2.0.7 everything works fine.

**Comment by philipvandebriel** [comment]
We have ,

Magento 2.1.0
Php 5.6
Redis
Debian 8.5 - 64G
Mysql 5.6.32

**Comment by pboisvert** [comment]
@PingusPepan 

Prepare >10 root categories
**Fill with many categories and subcategories**
Try to add or edit product

Can you be more specific in terms of how many categories and subcategories were added before this problem appears?

**Comment by PingusPepan** [comment]
I don't know exactly because it was a migration, so all the categories
appeared at once.

I've tried removing about ~400 old categories but it made almost no
difference. We still have over 4000 categories in total.

I also believe that this problem will behave differently computer from
computer because it is a pure js problem. Magento overloads the js with too
much data. When I checked the HTML source code of product edit page, there
was a line with all the JS data and it was extremely long.

Maybe it would be a better idea to load only root and direct root
subcategories by the onload js and then, after clicking on the category
arrow, load the subcategories by ajax.

Also something to think about is to display only root categories associated
with currently selected product websites and if websites change, update
categories list by ajax. Of course it would have needed to be switchable in
admin as many different people have many different needs.

**Comment by philipvandebriel** [comment]
Hi ,
I think the problem is the next file page-image-uploader.js , my browser freeze on this moment. See picture attachment.

![schermafdruk 2016-09-03 09 10 29](https://cloud.githubusercontent.com/assets/15958461/18223408/30fd977e-71b7-11e6-8d72-4c90398192b5.png)

**Comment by PingusPepan** [comment]
@philipvandebriel The real culprit in your case is `ui-select-optgroup.html` though. After it loaded, it started crunching data and that's why it stalled loading of another resources. Try manually editing the `ui-select-optgroup.html` file and make it empty. Then try loading the page again. For me, when the file was empty, it was working just fine (except that I was missing functionality provided by the file)

**Comment by philipvandebriel** [comment]
@PingusPepan Yes you are right, If i make the file empty then the product edit page load in 3sec but if we save the product then it's need again 53sec.

**Comment by philipvandebriel** [comment]
With this solution the Customizable Options will not work anymore. Can't add Customizable Options.

**Comment by PingusPepan** [comment]
@philipvandebriel Yes, because it is not a solution, it just points to the file which causes the issue.
The real problem lies in the over usage of javascript in the admin area though.

**Comment by philipvandebriel** [comment]
Will this be fix in magento 2.1.2 ? And when is the next update ?

**Comment by philipvandebriel** [comment]
Same problem in Magento 2.1.2 !

**Comment by MLUKSP** [comment]
Hi,
I'm having this exact same issue, doing anything with a product, it can take anywhere from 40 seconds to over 80 seconds, and looks to be due to lots of JS being called..  has anyone had any luck with resolving this?
I am still on 2.1 - has a recent upgrade resolved it for anyone?

Thanks

**Comment by PingusPepan** [comment]
@MLUKSP I had no time to look deeper into the issue.
There was a commit https://github.com/magento/magento2/commit/bebb2ceb000b9296f9b04ecd5424ef66cd723fbf which was supposed to fix the issue but as someone reported in the comment below it doesn't seem to work in Magento 2.1.2 although I personally haven't tested the commit.
This leads me to believe that it will work in Magento 2.2 perharps. At least that's what I'm eagerly waiting for.

**Comment by philipvandebriel** [comment]
Yes this fix will work, i changed the files in 2.1.2

**Comment by MLUKSP** [comment]
Ah fantastic - thanks guys

**Comment by benmarks** [comment]
Checking on this issue for you all, stay tuned.

**Comment by MLUKSP** [comment]
Thanks  @benmarks , looking forward to this being resolved.
I didn't get arond to commenting back here to say it still doesn't work after the 2.1.2 update.

I'm also not 100% certain that its not theme related, but as its only happening in the product area i'm thinking not.

**Comment by PingusPepan** [comment]
@benmarks Yesterday I applied commits https://github.com/magento/magento2/commit/bebb2ceb000b9296f9b04ecd5424ef66cd723fbf and https://github.com/magento/magento2/commit/2c0bdb10c28c32f2f37f724e64125238e61835b8 and it seems to work just fine for ~3500 categories.

If your internal testing confirms it works, can we hopefully see this released in 2.1.3?

**Comment by philipvandebriel** [comment]
Yes this will work, i changed in 2.1.2 with 4000 categories 3 weeks ago.

**Comment by PingusPepan** [comment]
@philipvandebriel If not already, apply also this commit: https://github.com/magento/magento2/commit/2c0bdb10c28c32f2f37f724e64125238e61835b8 it fixes a bug when if you hover over category, the search in categories does not work.

**Comment by philipvandebriel** [comment]
@PingusPepan Thanks!

**Comment by tkacheva** [comment]
Has to be addressed in this commit https://github.com/magento/magento2/commit/cfb571f6fb7dd59a4bb470a8ecbb60a3138833bc

**Comment by PingusPepan** [comment]
Looks like fixed in 2.1.3! Update and test if it is fixed for you

**Comment by philipvandebriel** [comment]
Just finish the update to 2.1.3, yes it is fixed.

**Comment by PingusPepan** [comment]
Fixed in 2.1.3, closing. Thanks to everyone who participated!
